internal/resources/aws/util.go

```diff
@@ -662,3 +670,26 @@ var InstanceTypeToVCPU = map[string]int64{
 	"z1d.metal":        48,
 	"z1d.xlarge":       4,
 }
+
+var reservedTermsMapping = map[string]string{
+	"1_year": "1yr",
+	"3_year": "3yr",
+}
+
+var reservedPaymentOptionMapping = map[string]string{
+	"no_upfront":      "No Upfront",
+	"partial_upfront": "Partial Upfront",
+	"all_upfront":     "All Upfront",
+}
+
+func validateReservedInstanceParams(reservedInstanceTerm, reservedInstancePaymentOption string) (bool, string) {
+	validTerms := sliceOfKeysFromMap(reservedTermsMapping)
+	if !stringInSlice(validTerms, reservedInstanceTerm) {
+		return false, fmt.Sprintf("Invalid reserved_instance_term, ignoring reserved options. Expected: 1_year, 3_year. Got: %s", reservedInstanceTerm)
+	}
+	validOptions := sliceOfKeysFromMap(reservedPaymentOptionMapping)
+	if !stringInSlice(validOptions, reservedInstancePaymentOption) {
+		return false, fmt.Sprintf("Invalid reserved_instance_payment_option, ignoring reserved options. Expected: no_upfront, partial_upfront, all_upfront. Got: %s", reservedInstancePaymentOption)
+	}
+	return true, ""
+}
```

Comment:
After poking around AWS pricing calculator I found another distinction from EC2 instance: `3_year` term doesn't have `no_upfront` value. If we allow to pass it, we'll end up with a warning "No products found". I think we can improve the validation here. I still think reusing it for EC2 instance is a great idea, so what do you think if we organize the code under a struct, so it is more seamless? Here's an example:
```go
type reservedInstanceResolver struct {
	term          string
	paymentOption string
	dbInstance    bool
}

func (r reservedInstanceResolver) Term() string {
	return reservedTermsMapping[r.term]
}

func (r reservedInstanceResolver) PaymentOption() string {
	return reservedPaymentOptionMapping[r.paymentOption]
}

func (r reservedInstanceResolver) validate() (bool, string) {
	validTerms := sliceOfKeysFromMap(reservedTermsMapping)
	if !stringInSlice(validTerms, r.term) {
		return false, fmt.Sprintf("Invalid reserved_instance_term, ignoring reserved options. Expected: %s. Got: %s", strings.Join(validTerms, ", "), r.term)
	}
	validOptions := sliceOfKeysFromMap(reservedPaymentOptionMapping)

	if r.term == "3_year" && r.dbInstance {
		validOptions = []string{"partial_upfront", "all_upfront"}
	}

	if !stringInSlice(validOptions, r.paymentOption) {
		return false, fmt.Sprintf("Invalid reserved_instance_payment_option, ignoring reserved options. Expected: %s. Got: %s", strings.Join(validOptions, ", "), r.paymentOption)
	}
	return true, ""
}

# EC2 instance
resolver := reservedInstanceResolver{
	term:          strVal(r.ReservedInstanceTerm),
	paymentOption: strVal(r.ReservedInstancePaymentOption),
}

# DB/RDS instance
resolver := reservedInstanceResolver{
	term:          strVal(r.ReservedInstanceTerm),
	paymentOption: strVal(r.ReservedInstancePaymentOption),
	dbInstance:    true,
}

# And then use it like
valid, err := resolver.validate()

# And hide the mapping maps with methods:
TermLength:         strPtr(resolver.Term()),
TermPurchaseOption: strPtr(resolver.PaymentOption()),
```

Reply:
Thank you @vdmgolub for the review! 

My bad I missed the UsageSchema items for RDSClusterInstance! Thank you for fixing it :)   

I also noticed too that 3_year term doesn't have no_upfront. In fact while updating the tests I got that warning but I thought this was not a problem because also terraform won't generate any plan in that case. So I considered it as an "impossibile" input.  But of course I agree that validation can be improved to avoid noisy warnings for that case. 

Great idea creating a new struct to handle and embed everything regarding reservation. This can be useful for all reservation cases, for example another case is the one related to OpenSearch reserved instances (I was thinking about opening an issue to implement this too). So we may can consider some other kind of mapping instead of several booleans like `dbInstance    bool` to differentiate all the services? Should this struct stay in util.go or we can evaluate to create a new file for it?

Reply:
No worries at all, team work 🤜 🤛 :) When making similar changes in several places it's easy to overlook things, talking from experience 😅.

You are correct, Terraform would complain and it is a good safety net. However with our new HCL parsing option using `--terraform-parse-hcl` flag (which, by the way, will become a default in the upcoming [v0.10](https://github.com/infracost/infracost/releases/tag/v0.10.0-beta.1) release) Terraform won't watch our backs anymore and we'll have to rely on our own validations. :)

I think we could keep it in utils for now to move this PR forward. If you consider to implement the OpenSearch changes, I think it will be a great opportunity to refactor it and decide what works better for that resource. How does it sound to you?

Reply:
Cool I did not know about `--terraform-parse-hcl` feature!

You are right, just keep it simple for the moment to move the this PR forward. I pushed the reservedInstanceResolver with the new validation in the latest commit!